typed-racket-lib/typed-racket/typecheck/tc-envops.rkt

```diff
@@ -31,33 +33,57 @@
        [props
         (let loop ([ps atoms]
                    [negs '()]
+                   [new '()]
                    [Γ (env-replace-props env props)])
           (match ps
             [(cons p ps)
              (match p
-               [(TypeProp: (Path: lo x) pt)
-                #:when (and (not (is-var-mutated? x))
-                            (identifier-binding x))
-                (let* ([t (lookup-type/lexical x Γ #:fail (λ (_) Univ))]
-                       [new-t (update t pt #t lo)])
-                  (and (not (Bottom? new-t))
-                       (loop ps negs (env-set-type Γ x new-t))))]
+               [(TypeProp: (Path: pes (? identifier? x)) pt)
```

Comment:
most types are stored w/ their associated identifier.

Occasionally, however, we want to record the type of something w/o an underlying identifier (i.e. it has a DeBruijn index, its a linear expression, etc), and so when this occurs we store the type of the whole Object in the type environment.

It might be nice to make the lexical environment ONLY talk about the types of objects (and not identifies), but that would be a larger refactoring... which I remember considering... and I can't remember why I didn't just do it.